cluster-up/cluster/kind-k8s-sriov-1.17.0/config_sriov.sh

```diff
@@ -25,79 +22,127 @@ SRIOVDP_RESOURCE_NAME="sriov_net"
-    pf_array+=("$pf_name")
-    [ "${#pf_array[@]}" -eq "$pf_count_per_node" ] && break
+    pfs_names+=( $pf_name )
   done
 
-  [ "${#pf_array[@]}" -lt "$pf_count_per_node" ] && \
-    echo "FATAL: Not enough PFs allocated, PF_BLACKLIST (${PF_BLACKLIST[@]}), PF_COUNT_PER_NODE ${PF_COUNT_PER_NODE}" >&2 && \
-    return 1
+  echo "${pfs_names[@]}"
+}
+
+function validate_nodes_sriov_allocatable_resource() {
+  local -r resource_name="$SRIOVDP_RESOURCE_PREFIX/$SRIOVDP_RESOURCE_NAME"
+  local -r sriov_nodes=$(_kubectl get nodes -l $SRIOV_NODE_LABEL -o custom-columns=:.metadata.name --no-headers)
 
-  echo "${pf_array[@]}"
+  for sriov_node in $sriov_nodes; do
+    num_vfs=$(total_vfs_count_on_node "$sriov_node")
+    sriov_components::wait_allocatable_resource "$sriov_node" "$resource_name" "$num_vfs"
+  done
 }
 
-# The first worker needs to be handled specially as it has no ending number, and sort will not work
-# We add the 0 to it and we remove it if it's the candidate worker
-WORKER=$(_kubectl get nodes | grep $WORKER_NODE_ROOT | sed "s/\b$WORKER_NODE_ROOT\b/${WORKER_NODE_ROOT}0/g" | sort -r | awk 'NR==1 {print $1}')
-if [[ -z "$WORKER" ]]; then
-  SRIOV_NODE=$MASTER_NODE
-else
-  SRIOV_NODE=$WORKER
-fi
+function total_vfs_count_on_node() {
+  local -r node_name=$1
+  local -r node_pid=$(docker inspect -f '{{.State.Pid}}' "$node_name")
+  local -r pfs_sriov_numvfs=( $(cat /proc/$node_pid/root/sys/class/net/*/device/sriov_numvfs) )
+  local total_vfs_on_node=0
+
+  for num_vfs in "${pfs_sriov_numvfs[@]}"; do
+    total_vfs_on_node=$((total_vfs_on_node + num_vfs))
+  done
+
+  echo "$total_vfs_on_node"
+}
 
-# this is to remove the ending 0 in case the candidate worker is the first one
-if [[ "$SRIOV_NODE" == "${WORKER_NODE_ROOT}0" ]]; then
-  SRIOV_NODE=${WORKER_NODE_ROOT}
-fi
+worker_nodes=($(_kubectl get nodes -l node-role.kubernetes.io/worker -o custom-columns=:.metadata.name --no-headers))
+nodes_count=${#worker_nodes[@]}
+[ "$nodes_count" -eq 0 ] && echo "FATAL: no worker nodes found" >&2 && exit 1
 
-NODE_PFS=($(move_sriov_pfs_netns_to_node "$SRIOV_NODE" "$PF_COUNT_PER_NODE"))
-NODE_PF=$NODE_PFS
+pfs_names=( $(get_pfs_names) )
+pf_count="${#pfs_names[@]}"
+[ "$pf_count" -le 0 ] && echo "FATAL: Could not find available sriov PF's" >&2 && exit 1
 
-SRIOV_NODE_CMD="docker exec ${SRIOV_NODE}"
-NODE_PF_NUM_VFS=$(${SRIOV_NODE_CMD} cat /sys/class/net/$NODE_PF/device/sriov_totalvfs)
+total_pf_required=$((nodes_count*NODES_PFS_PER_NODE))
+[ "$pf_count" -lt $total_pf_required ] && \
+  echo "FATAL: there are not enough PF's on the host, try to reduce NODES_PFS_PER_NODE
+  Worker nodes count: $nodes_count
+  PF per node count:  $NODES_PFS_PER_NODE
+  Total PF count required:  $total_pf_required" >2 && return 1
 
-# KIND mounts sysfs as readonly, this script requires R/W access to sysfs
-${SRIOV_NODE_CMD} mount -o remount,rw /sys
-${SRIOV_NODE_CMD} chmod 666 /dev/vfio/vfio
+## Move SRIOV Physical Functions to worker nodes and configure VF's
```

Comment:
Can this be a function?

Reply:
I tried few things and yes its possible, I will change it